File: maga_transformer/async_decoder_engine/generate_stream.py
import copy
import torch
import logging
import numpy as np
from enum import Enum
from typing import Any, List, Optional
from threading import Lock
from pydantic import BaseModel, Field, PrivateAttr

from maga_transformer.utils.time_util import current_time_ms
from maga_transformer.utils.stop_utils import create_stop_criteria_list
from maga_transformer.metrics import kmonitor, GaugeMetrics
from maga_transformer.models.base_model import GenerateInput, GenerateOutput
from maga_transformer.async_decoder_engine.ptuning.ptuning import PrefixInfo
from maga_transformer.utils.util import AtomicCounter

class Status(Enum):
    WAITING = 0
    RUNNING = 1
    STOPPED = 2
    FINISHED = 3

class GenerateStatus(BaseModel):
    status: Status = Status.WAITING
    error_info: str = ''

stream_counter = AtomicCounter()

class GenerateStream(BaseModel):
    _input: GenerateInput
    _status: GenerateStatus = PrivateAttr(default_factory=GenerateStatus)
    _output: GenerateOutput = PrivateAttr(default_factory=GenerateOutput)
    _complete_token_ids: torch.Tensor
    _max_seq_len: int = PrivateAttr(default_factory=int)
    _block_indice : List[List[int]] = PrivateAttr(default_factory=list)
    _reuse_length : int = PrivateAttr(default_factory=int)
    _resource_dtors: List[Any] = []
    medusa_state: Any = None
    _stream_id: int = PrivateAttr(default_factory=int)
    _released: bool = PrivateAttr(default_factory=bool)

    def __init__(self, input: GenerateInput, max_seq_len: int=2048):
        super().__init__()
        self._input = input
        self._max_seq_len = max_seq_len
        self._seq_length = self.input_length
        self._begin_time = current_time_ms()
        self._resource_dtors = []

        self._output.output_ids = torch.zeros([self._input.generate_config.num_beams, 0], dtype=torch.int32)
        self._complete_token_ids = torch.zeros([self._input.generate_config.num_beams, max_seq_len], dtype=torch.int32)
        self._complete_token_ids[0, :self._seq_length] = self._input.token_ids
        self._cum_log_probs = torch.zeros([self._input.generate_config.num_beams], dtype=torch.float32)
        self._stop_criteria_list = create_stop_criteria_list(
            self._input.generate_config.stop_words_list,
            self._input.generate_config.stop_words_str,
            self._input.tokenizer)
        self._ptuning_info = PrefixInfo()
        self._lock = Lock()

        global stream_counter
        self._stream_id = stream_counter.increment()
        self._released = False
        # for mulit_task_prompt
        self._require_release = True

    def add_resource_dtor(self, dtor: Any):
        self._resource_dtors.append(dtor)
        
    def set_require_release(self, require_release: bool):
        self._require_release = require_release
        
    def clear_resource_dtor(self):
        self._resource_dtors.clear()

    def release_resource(self):        
        if self._released or not self._require_release:
            return
        for dtor in self._resource_dtors:
            dtor()
        self._resource_dtors.clear()
        self._released = True

    def update_prefix(self, ptuning_info: PrefixInfo):
        self._ptuning_info = ptuning_info
        if ptuning_info.prefix_tensors is not None:
            self._input.update_prefix(ptuning_info.prefix_tensors)
            self._seq_length = self.input_length
            self._complete_token_ids[0, :self._seq_length] = self._input.token_ids

    @property
    def ptuning_info(self):
        return self._ptuning_info

    @property
    def cum_log_probs(self):
        return self._cum_log_probs

    @property
    def complete_token_ids(self):
        return self._complete_token_ids[:,:self.seq_length]

    @property
    def input_token_ids(self):
        return self._complete_token_ids[:, :self.input_length]

    @property
    def reuse_length(self):
        return self._reuse_length

    @property
    def input_length(self):
        return self._input.input_length

    @property
    def prefix_length(self):
        return self._input.prefix_length

    @property
    def seq_length(self):
        return self._seq_length

    @property
    def generate_config(self):
        return self._input.generate_config

    @property
    def images(self):
        return self._input.images

    @property
    def lora_id(self):
        return self._input.lora_id

    @property
    def output(self):
        with self._lock:
            output = copy.copy(self._output)
            output.aux_info = copy.copy(output.aux_info)
            return output

    def set_stop(self, err: str):
        with self._lock:
            self._status.status = Status.STOPPED
            self._status.error_info = err

    def stop_and_release(self, err: str):
        self.set_stop(err)
        self.release_resource()

    def set_running(self):
        with self._lock:
            if self._stopped:
                return False
            self._report_wait_time()
            self._status.status = Status.RUNNING
            return True

    def _set_finished(self):
        self._status.status = Status.FINISHED

    @property
    def _stopped(self):
        return self._status.status == Status.STOPPED

    @property
    def stopped(self):
        with self._lock:
            return self._status.status == Status.STOPPED

    @property
    def stop_reason(self):
        with self._lock:
            return self._status.error_info

    @property
    def finished(self):
        with self._lock:
            return self._status.status == Status.FINISHED

    def check_timeout(self):
        running_time = current_time_ms() - self._begin_time
        if self.generate_config.timeout_ms > 0 and self.generate_config.timeout_ms < running_time:
            self.stop_and_release(f"query has been running {running_time} ms, it's timeout")

    def _report_wait_time(self):
        kmonitor.report(GaugeMetrics.ASYNC_WAIT_WAIT_TIME_METRIC, current_time_ms() - self._begin_time)

    def _report_first_token_rt(self):
        kmonitor.report(GaugeMetrics.FT_FIRST_TOKEN_RT_METRIC, current_time_ms() - self._begin_time)

    def set_kvcache(self, block_indice: List[List[int]], reuse_length: int):
        self._block_indice = block_indice
        self._reuse_length = reuse_length

    def update(self,
               new_tokens: torch.Tensor,
               num_new_tokens: int,
               finished: bool,
               hidden_states: Optional[torch.Tensor],
               logits: Optional[torch.Tensor],
               cum_log_probs: Optional[torch.Tensor]):
        with self._lock:
            if self._output.aux_info.iter_count == 0:
                self._report_first_token_rt()

            # NOTE: new tokens indicate num of newly genearted tokens
            # typically 1 but can be > 1 under speculative decoding
            # This differs from new_tokens.shape[-1] under beam search case,
            # which needs to update all the generated tokens each update.
            update_length = new_tokens.shape[-1]
            update_to_pos = self._seq_length + num_new_tokens
            update_from_pos = update_to_pos - update_length
            self._complete_token_ids[:, update_from_pos:update_to_pos] = new_tokens
            for i in range(num_new_tokens):
                self._seq_length += 1
                if self._need_finish():
                    finished = True
                    break
            if finished:
                self._set_finished()
            self._output.output_ids = self._complete_token_ids[:,self.input_length:self._seq_length]
            self._output.input_ids = self.input_token_ids
            self._output.hidden_states = hidden_states
            if len(self._input.generate_config.select_tokens_id) > 0:
                self._output.logits = logits[:,self._input.generate_config.select_tokens_id]
            else:
                self._output.logits = logits

            self._output.finished = finished
            self._output.aux_info.cost_time = current_time_ms() - self._begin_time
            self._output.aux_info.input_len = self._input.prompt_length
            self._output.aux_info.prefix_len = self._input.prefix_length
            self._output.aux_info.output_len = self._output.output_ids.shape[-1]
            self._output.aux_info.cum_log_probs = cum_log_probs.tolist() if cum_log_probs is not None else None
            self._output.aux_info.iter_count += 1
            self._output.aux_info.reuse_len = self._reuse_length

    def set_loss(self, loss):
        self._output.loss = loss

    def add_block_index(self, block_index: List[List[int]]):
        assert len(block_index) == len(self._block_indice)
        for i in range(len(block_index)):
            self._block_indice[i].extend(block_index[i])

    def pop_block_indice(self):
        block_indice = self._block_indice
        self._block_indice = [[]]
        return block_indice

    @property
    def block_indice(self):
        return self._block_indice

    def _need_finish(self):
        return self._seq_length >= min(self._max_seq_len, self.generate_config.max_new_tokens + self._input.input_length) \
            or (self._seq_length >= self.generate_config.min_new_tokens + self._input.input_length and self._invoke_stop_words_criterion())

    def _invoke_stop_words_criterion(self):
        if self._seq_length == self.input_length:
            return False
        tokens_to_check = self._complete_token_ids[0,self.input_length:self._seq_length].tolist()
        for stop_criteria in self._stop_criteria_list:
            if stop_criteria(tokens_to_check):
                return True
        return False

    def __hash__(self):
        return id(self)

    def __eq__(self, other):
        return id(self) == id(other)

File: maga_transformer/cpp/dataclass/GenerateStream.cc
#include <cstddef>
#include <memory>
#include "autil/EnvUtil.h"
#include "maga_transformer/cpp/dataclass/GenerateStream.h"
#include "maga_transformer/cpp/dataclass/Query.h"
#include "src/fastertransformer/core/Buffer.h"
#include "src/fastertransformer/core/Types.h"
#include "src/fastertransformer/devices/DeviceFactory.h"
#include "maga_transformer/cpp/metrics/RtpLLMMetrics.h"
#include "src/fastertransformer/th_op/GptInitParameter.h"
#include "src/fastertransformer/core/torch_utils/BufferTorchUtils.h"
#include "src/fastertransformer/utils/assert_utils.h"

using namespace std;

namespace rtp_llm {

GenerateStream::GenerateStream(const shared_ptr<GenerateInput>& input,
                               const ft::GptInitParameter&      params,
                               const ResourceContext&           resource_context,
                               kmonitor::MetricsReporterPtr     metrics_reporter)
    : generate_input_(input)
    , max_seq_len_(params.max_seq_len_)
    , vocab_size_(params.vocab_size_)
    , stream_cache_resource_(this, resource_context, input->need_release_resource)
    , need_release_resource_(input->need_release_resource)
    , enable_fast_gen_(params.enable_fast_gen_)
    , metrics_reporter_(metrics_reporter)
    , special_tokens_(params.special_tokens_)
    , output_mutex_(std::make_shared<std::mutex>())
    , generate_outputs_queue_(std::make_shared<autil::SynchronizedQueue<GenerateOutputs>>())

{
    updatePrefix(resource_context.system_prompt);
    seq_length_ = generate_input_->inputLength();
    start_check_seq_length_ = seq_length_;
    last_output_pos_ = seq_length_;
    common_len_ = seq_length_;
    max_chunk_len_ = seq_length_;

    begin_time_us_      = input->begin_time_ms;
    device_             = ft::DeviceFactory::getDefaultDevice();
    complete_token_ids_ = device_->allocateBuffer(
        {ft::DataType::TYPE_INT32, {(size_t)tileNum(), (size_t)max_seq_len_}, ft::AllocationType::HOST}, {});
    if (generate_input_->generate_config->calculate_loss && inputLength() > 1) {
        loss_ = device_->allocateBuffer(
                {ft::DataType::TYPE_FP32, {(size_t)inputLength() - 1}, ft::AllocationType::HOST}, {});
    }
    memset(complete_token_ids_->data(), 0, complete_token_ids_->sizeBytes());
    for (int i = 0; i < tileNum(); ++i) {
        memcpy(complete_token_ids_->dataWithOffset<int32_t>(i * max_seq_len_),
               generate_input_->input_ids->data(),
               generate_input_->input_ids->sizeBytes());
    }

    cum_log_probs_ =
        device_->allocateBuffer({ft::DataType::TYPE_FP32, {(size_t)tileNum()}, ft::AllocationType::HOST}, {});
    memset(cum_log_probs_->data(), 0, cum_log_probs_->sizeBytes());


    generate_status_.status = GenerateState::WAITING;
    sub_generate_status_.clear();
    sub_generate_status_.resize(tileNum());
    for (int i = 0; i < tileNum(); ++i) {
        sub_generate_status_[i].status = GenerateState::WAITING;
    }

    stream_cache_resource_.init(tileNum());

    perf_test_ = autil::EnvUtil::getEnv("PERF_TEST", false);
    // TODO: need fix context block copy
    perf_test_ = true;
}

absl::StatusOr<int> GenerateStream::acquireCapacity(int token_capacity) {
    if (token_capacity <= 0) {
        return absl::InternalError("token_capacity is <= 0");
    }
    if (isChunkStream()) {
        // TODO(xinfei.sxf) add min_chunk_len ?
        if (current_chunk_len_ == 0) {
            current_chunk_len_ = reuse_length_;
        }
        auto remaining_token = max_chunk_len_ - current_chunk_len_;
        last_chunk_len_ = current_chunk_len_;
        if (token_capacity > remaining_token) {
            current_chunk_len_ = max_chunk_len_;
            return remaining_token;
        } else {
            current_chunk_len_ += token_capacity;
            return token_capacity;
        }
    } else if (!isContextStream()) {
        return 1;
    }
    FT_CHECK(false);
    return absl::InternalError("unexpected call");
}

void GenerateStream::cancel() {
    cancelled_ = true;
    setStop("cancel stream");
}

absl::StatusOr<int> GenerateStream::initKVBlock(int token_capacity, size_t reserve_step) {
    if (generate_status_.status == GenerateState::WAITING) {
        wait_time_us_ = autil::TimeUtility::currentTimeInMicroSeconds() - begin_time_us_;
    } else if (generate_status_.status == GenerateState::PAUSED) {
        pause_time_us_ += autil::TimeUtility::currentTimeInMicroSeconds() - last_pause_us_;
    }
    return stream_cache_resource_.initKVBlock(token_capacity, reserve_step);
}

absl::StatusOr<int>GenerateStream::incrKVBlock(int token_capacity, size_t reserve_step) {
    return stream_cache_resource_.incrKVBlock(token_capacity, reserve_step);
}

int GenerateStream::tryReleaseKVBlock(int nums) {
    auto release_blocks = stream_cache_resource_.tryReleaseKVBlock(nums);
    incrFallbackBlock(release_blocks);
    return release_blocks;
}

void GenerateStream::releaseResource() {
    if (need_release_resource_) {
        stream_cache_resource_.releaseResource();
    }
}
int GenerateStream::nextNeedBlockNums(size_t reserve_step) const {
    // TODO: maybe need fix when context and reuse
    return stream_cache_resource_.singleBatchNeedBlocks(seq_length_ + reserve_step) * batchSize();
}

void GenerateStream::incrFallbackBlock(int fallback_blocks) {
    fallback_blocks_ += fallback_blocks;
    fallback_times_ += 1;
}

std::shared_ptr<GenerateInput> GenerateStream::generateInput() const {
    return generate_input_;
}
std::shared_ptr<GenerateConfig>& GenerateStream::generateConfig() const {
    return generate_input_->generate_config;
}
bool GenerateStream::isStreaming() const {
    return generate_input_->generate_config->is_streaming;
}
int64_t GenerateStream::streamId() const {
    return generate_input_->request_id;
}
int GenerateStream::loraId() const {
    return generate_input_->lora_id;
}

std::string GenerateStream::adapterName() const {
    return generate_input_->generate_config->adapter_name;
}

ft::SpecialTokens GenerateStream::specialTokens() const {
    return special_tokens_;
}

int GenerateStream::tileNum() const {
    return std::max(numBeams(), numReturnSequences());
}

int GenerateStream::batchSize() const {
    return seq_length_ == inputLength() && !perf_test_ ? 1 : tileNum();
}

int GenerateStream::numBeams() const {
    return generate_input_->generate_config->num_beams;
}

int GenerateStream::numReturnSequences() const {
    return generate_input_->generate_config->num_return_sequences;
}

bool GenerateStream::calculateLoss() const {
    return loss_ && loss_index_ < inputLength() - 1;
}

void GenerateStream::updatePrefix(const std::shared_ptr<SystemPrompt>& system_prompt) {
    if (system_prompt) {
        prompt_param_ = system_prompt->getPromptParams(*generate_input_->generate_config);
        if (!prompt_param_.prompt_token.empty()) {
            generate_input_->updatePrefix(prompt_param_.prompt_token);
        }
    }
}

size_t GenerateStream::maxSeqLen() const {
    return max_seq_len_;
}

int GenerateStream::inputLength() const {
    return generate_input_->inputLength();
}

int GenerateStream::currentChunkLen() const {
    return current_chunk_len_;
}

void GenerateStream::resetChunkLen(int chunk_len, int max_chunk_len) {
    last_chunk_len_ = 0;
    current_chunk_len_ = chunk_len;
    max_chunk_len_ = max_chunk_len;
}

int GenerateStream::seqLength() const {
    return seq_length_;
}

int GenerateStream::commonLen() const {
    return common_len_;
}

int GenerateStream::adjustedCommonLen() const {
    return tileNum() == 1 ? seq_length_ : inputLength() / seqSizePerBlock() * seqSizePerBlock();
}

int GenerateStream::seqSizePerBlock() const {
    return stream_cache_resource_.seqSizePerBlock();
}

int GenerateStream::contextLength() const {
    int begin_pos = prefixLength();
    int end_pos = isChunkStream() ? currentChunkLen() : seq_length_;
    return end_pos - begin_pos;
}

int GenerateStream::inputPrefixLength() const {
    return generate_input_->prefix_length;
}

int GenerateStream::prefixLength() const {
    if (fallback_prefix_length_) {
        return fallback_prefix_length_;
    } else if (last_chunk_len_) {
        return last_chunk_len_;
    }
    return reuse_length_;
}

int GenerateStream::reuseLength() const {
    if (multimodalFeatures().has_value()) {
        // prompt with multimodal input cannot use reuse cache for now
        return 0;
    }
    return reuse_length_;
}

void GenerateStream::setReuseLength(int reuse_length) {
    reuse_length_ = reuse_length;
}

int GenerateStream::fallbackPrefixLength() const {
    return fallback_prefix_length_;
}

void GenerateStream::setFallbackPrefixLength(int fallback_prefix_length) {
    fallback_prefix_length_ = fallback_prefix_length;
}

bool GenerateStream::isContextStream() const {
    return is_context_stream_;
}

bool GenerateStream::isChunkStream() const {
    return enable_fast_gen_ && current_chunk_len_ < max_chunk_len_;
}

const ft::BufferPtr& GenerateStream::cumLogProbs() const {
    return cum_log_probs_;
}

const ft::BufferPtr& GenerateStream::completeTokenIds() {
    return complete_token_ids_;
}

std::vector<int> GenerateStream::completeTokenIdsVec(int batch_idx) {
    FT_CHECK(batch_idx < tileNum());
    return fastertransformer::buffer2vector<int>(complete_token_ids_->view(batch_idx, 1), seq_length_);
}

std::vector<int> GenerateStream::commonCompleteTokenIdsVec(int batch_idx) {
    FT_CHECK(batch_idx < tileNum());
    return fastertransformer::buffer2vector<int>(complete_token_ids_->view(batch_idx, 1), common_len_);
}

int GenerateStream::currentExecuteTokenSize() {
    return currentExecuteTokens(0).size() * batchSize();
}

vector<int> GenerateStream::contextTokens(int batch_idx) const {
    return fastertransformer::buffer2vector<int>(
            (*complete_token_ids_)[batch_idx].view(prefixLength(), contextLength()));
}

std::optional<std::vector<torch::Tensor>>& GenerateStream::multimodalFeatures() const {
    return generate_input_->multimodal_features;
}

int GenerateStream::multimodalFeaturesLength() const {
    auto& features = multimodalFeatures();
    if (features) {
        return features.value().size() * batchSize();
    }
    return 0;
}

std::optional<ft::BufferPtr>& GenerateStream::multimodalLocations() const {
    return generate_input_->mm_locs;
}

vector<int> GenerateStream::textTokensMask() const {
    if (!generate_input_->text_tokens_mask) {
        return {};
    }
    auto token_masks = fastertransformer::buffer2vector<int>(*generate_input_->text_tokens_mask.value());
    if (reuseLength() > 0) {
        return vector<int>(token_masks.begin() + reuseLength(), token_masks.end());
    } else {
        return token_masks;
    }
}

vector<int> GenerateStream::currentExecuteTokens(int batch_idx) const {
    // TODO(xinfei.sxf) 在query部分回退，重运行case下，这个不对
    if (isContextStream()) {
        return contextTokens(batch_idx);
    } else {
        return {*(*complete_token_ids_)[batch_idx].dataWithOffset<int>(seq_length_ - 1)};
    }
}

void GenerateStream::step() {
    // iter_count represents the times of the stream participates in running
    iter_count_++;
    if (isContextStream()) {
        setFallbackPrefixLength(0);
    }
}

void GenerateStream::checkTimeout() {
    auto running_time_ms = (autil::TimeUtility::currentTimeInMicroSeconds() - begin_time_us_) / 1000;
    auto timeout_ms      = generate_input_->generate_config->timeout_ms;
    if (timeout_ms > 0 && timeout_ms < running_time_ms) {
        stopAndRelease("query has been running " + std::to_string(running_time_ms) + " ms, "
                       + "timeout_ms = " + std::to_string(timeout_ms) + ", it's timeout",
                       absl::StatusCode::kDeadlineExceeded);
    }
}

bool GenerateStream::checkTokenId(int token_id) {
    return token_id >= 0 && token_id < vocab_size_;
}

void GenerateStream::setStop(const std::string& err_msg, absl::StatusCode err_code) {
    std::lock_guard<std::mutex> lock(*output_mutex_);
    FT_LOG_WARNING("stop stream: %d %s", streamId(), err_msg.c_str());
    generate_status_.status     = GenerateState::STOPPED;
    generate_status_.error_code = err_code;
    generate_status_.error_info = err_msg;
}

void GenerateStream::stopAndRelease(const std::string& err_msg, absl::StatusCode err_code) {
    setStop(err_msg, err_code);
    releaseResource();
}

void GenerateStream::setPaused() {
    std::lock_guard<std::mutex> lock(*output_mutex_);
    if (stoppedWithoutLock()) {
        return;
    }
    is_context_stream_      = true;
    generate_status_.status = GenerateState::PAUSED;
    last_pause_us_          = autil::TimeUtility::currentTimeInMicroSeconds();
}

bool GenerateStream::setRunning() {
    std::lock_guard<std::mutex> lock(*output_mutex_);
    if (stoppedWithoutLock()) {
        return false;
    }
    generate_status_.status = GenerateState::RUNNING;
    return true;
}

void GenerateStream::setFinishedWithoutLock() {
    generate_status_.status = GenerateState::FINISHED;
    for (int i = 0; i < tileNum(); ++i) {
        sub_generate_status_[i].status = GenerateState::FINISHED;
    }
}

bool GenerateStream::stoppedWithoutLock() {
    return generate_status_.status == GenerateState::STOPPED;
}

bool GenerateStream::stopped() {
    std::lock_guard<std::mutex> lock(*output_mutex_);
    return generate_status_.status == GenerateState::STOPPED;
}

bool GenerateStream::paused() {
    std::lock_guard<std::mutex> lock(*output_mutex_);
    return generate_status_.status == GenerateState::PAUSED;
}

std::string GenerateStream::stopReason() {
    std::lock_guard<std::mutex> lock(*output_mutex_);
    return generate_status_.error_info;
}

bool GenerateStream::finished() {
    std::lock_guard<std::mutex> lock(*output_mutex_);
    return generate_status_.status == GenerateState::FINISHED;
}

size_t GenerateStream::iterCount() const {
    return iter_count_;
}

void GenerateStream::setKVCache(const BatchKVCacheBlockAddr& kv_cache_block_addr) {
    stream_cache_resource_.setKVCache(kv_cache_block_addr);
}

const BatchKVCacheBlockAddr& GenerateStream::kvCache() const {
    return stream_cache_resource_.kvCache();
}

const ResourceContext& GenerateStream::resourceContext() const {
    return stream_cache_resource_.resourceContext();
}

size_t GenerateStream::maxBlockSize() const {
    return stream_cache_resource_.maxBlockSize();
}

bool GenerateStream::needFinish() {
    return seq_length_ >= std::min(max_seq_len_,
                                   generate_input_->generate_config->max_new_tokens + generate_input_->inputLength())
           || needFinishBySPTokens();
}

bool GenerateStream::needFinishBySPTokens() {
    matchEosToken();
    matchStopWordsList();
    // num beams, finished by batch 0
    if (numBeams() != 1) {
        return sub_generate_status_[0].status == GenerateState::FINISHED;
    }
    // num sequence, finished by all batch
    return std::all_of(sub_generate_status_.begin(), sub_generate_status_.end(), [](GenerateStatus& generate_status) {
        return generate_status.status == GenerateState::FINISHED;
    });
}

void GenerateStream::matchEosToken() {
    for (int i = 0; i < tileNum(); ++i) {
        matchEosToken(i);
    }
}

void GenerateStream::matchEosToken(int batch_id) {
    int* token_ids_ = (int*)complete_token_ids_->view(batch_id, 1).data();
    for (size_t i = start_check_seq_length_; i <= seq_length_; ++i) {
        if (special_tokens_.eos_token_id_ == token_ids_[i - 1]) {
            sub_generate_status_[batch_id].status = GenerateState::FINISHED;
            seq_length_ = i;
        }
    }
}
void GenerateStream::matchStopWordsList() {
    if (seq_length_ < generate_input_->generate_config->min_new_tokens + inputLength()) {
        return;
    }
    if (seq_length_ == inputLength()) {
        return;
    }
    for (int i = 0; i < tileNum(); ++i) {
        matchStopWordsList(i);
    }
}

void GenerateStream::matchStopWordsList(int batch_id) {
    int* token_ids_ = (int*)complete_token_ids_->view(batch_id, 1).data();
    // note: stop_words_list in generate_config contains stop_words_list in special_tokens
    bool match = false;
    for (auto& stop_words : generate_input_->generate_config->stop_words_list) {
        for (size_t i = start_check_seq_length_; i <= seq_length_; ++i) {
            bool match_one   = true;
            size_t begin_index = i - stop_words.size();
            for (auto& token : stop_words) {
                if (token != token_ids_[begin_index++]) {
                    match_one = false;
                    break;
                }
            }

            if (match_one) {
                match = match_one;
                seq_length_ = i;
                break;
            }
        }

        if (match) {
            break;
        }
    }
    if (match) {
        sub_generate_status_[batch_id].status = GenerateState::FINISHED;
    }
}

void GenerateStream::update(const ft::BufferPtr&    new_tokens,
                            int               num_new_tokens,
                            const ft::BufferPtr& hidden_states,
                            const ft::BufferPtr& logits,
                            const ft::BufferPtr& cum_log_probs,
                            const ft::BufferPtr& all_probs) {
    FT_LOG_DEBUG(__PRETTY_FUNCTION__);
    is_context_stream_ = false;
    if (stoppedWithoutLock()) {
        return;
    }
    if (seq_length_ == generate_input_->inputLength()) {
        first_token_time_us_ = autil::TimeUtility::currentTimeInMicroSeconds() - begin_time_us_;
    }
    // # NOTE: new tokens indicate num of newly genearted tokens
    // # typically 1 but can be > 1 under speculative decoding
    // # This differs from new_tokens.shape[-1] under beam search case,
    // # which needs to update all the generated tokens each update.
    FT_CHECK(new_tokens->dim() == 2);
    for (size_t i = 0; i < tileNum(); ++i) {
        for (size_t j = 0; j < num_new_tokens; ++j) {
            auto current_token_id = *(*new_tokens)[i].dataWithOffset<int>(j);
            if (!checkTokenId(current_token_id)) {
                setStop("output token id:" + to_string(current_token_id) + " out of vocab size: " + to_string(vocab_size_));
                return;
            }
        }
        
        device_->copy({(*complete_token_ids_)[i].view(seq_length_, num_new_tokens), (*new_tokens)[i].view(0, num_new_tokens)});
    }
    setSeqLength(seq_length_ + num_new_tokens);

    updateOutput(new_tokens, hidden_states, logits, cum_log_probs, all_probs);
}

void GenerateStream::setLoss(const ft::Buffer& loss) {
    FT_CHECK(loss_index_ + loss.size() < inputLength());
    device_->copy({loss_->view(loss_index_, loss.size()), loss});
    loss_index_ += loss.size();
}

ft::BufferPtr GenerateStream::getLoss() {
    return loss_;
}

void GenerateStream::setMetricsReporter(kmonitor::MetricsReporterPtr metrics_reporter) {
    metrics_reporter_ = metrics_reporter;
}
void GenerateStream::reportMetric() {
    if (metrics_reporter_) {
        RtpLLMStreamMetricsCollector collector;
        collector.qps        = finished() || cancelled_;
        collector.cancel_qps = cancelled_;
        collector.error_qps  = stopped() && !cancelled_;
        if (finished() || cancelled_) {
            collector.reuse_length           = reuse_length_;
            collector.input_token_length     = inputLength();
            collector.output_token_length    = seq_length_ - generate_input_->inputLength();
            collector.iterate_count          = iter_count_;
            collector.query_batch_size       = tileNum();
            collector.total_latency_us       = autil::TimeUtility::currentTimeInMicroSeconds() - begin_time_us_;
            collector.first_token_latency_us = first_token_time_us_;
            collector.wait_latency_us        = wait_time_us_;
            collector.pause_latency_us       = pause_time_us_;
            collector.fallback_tokens        = fallback_blocks_ * seqSizePerBlock();
            collector.fallback_times         = fallback_times_;
        }
        metrics_reporter_->report<RtpLLMStreamMetrics, RtpLLMStreamMetricsCollector>(nullptr, &collector);
    }
}

std::string GenerateStream::debugString() const {
    std::stringstream debug_string;
    debug_string << "GenerateStream {"
                 << "generate_input:" << generate_input_->debugString() << ", max_seq_len:" << max_seq_len_
                 << ", input_length:" << inputLength() << ", seq_length:" << seq_length_
                 << ", reuse_length:" << reuse_length_ << ", current_chunk_len:" << current_chunk_len_
                 << ", last_chunk_len_:" << last_chunk_len_ << ", max_chunk_len_:" << max_chunk_len_
                 << ", batch_size:" << batchSize()
                 << ", tile_num:" << tileNum()
                 << ", need_release_resource: " << need_release_resource_
                 << ", fallback_prefix_length: " << fallback_prefix_length_;

    debug_string << ", complete_token_ids: [";
    for (size_t i = 0; i < tileNum(); i++) {
        debug_string << (*complete_token_ids_)[i].view(0, seq_length_).debugStringWithData<int32_t>() << ",";
    }

    debug_string << ", cum_log_probs: " << cum_log_probs_->debugStringWithData<float>();
    debug_string << ", stream_cache_resource: "<< stream_cache_resource_.debugString();
         
    debug_string << "}";
    return debug_string.str();
}

void GenerateStream::resetCommonLen() {
    if (tileNum() == 1) {
        common_len_ = seq_length_;
    }
}

void GenerateStream::setSeqLength(int seq_length) {
    if (seq_length > seq_length_) {
        start_check_seq_length_ = seq_length_ + 1;
    } else {
        start_check_seq_length_ = seq_length;
    }
    seq_length_ = seq_length;
    resetCommonLen();
}

void GenerateStream::setPerfTest(bool perf_test) {
    perf_test_ = perf_test;
}

void GenerateStream::setIsContextStream(bool is_context_stream) {
    is_context_stream_ = is_context_stream;
}

StreamCacheResource& GenerateStream::streamCacheResource() {
    return stream_cache_resource_;
}

void GenerateStream::CopyOnWrite(const GenerateStream& other_stream) {
    complete_token_ids_ = device_->clone({*other_stream.complete_token_ids_, ft::AllocationType::HOST});
    cum_log_probs_ = device_->clone({*other_stream.cum_log_probs_, ft::AllocationType::HOST});
    if (other_stream.calculateLoss()) {
        loss_ = device_->clone({*other_stream.loss_, ft::AllocationType::HOST});
    }
    stream_cache_resource_.setStream(this);
}

}  // namespace rtp_llm
